### PR TITLE
Bidirectional var operators

### LINKED ISSUE #93: Bidirectional Var Operators
If I invest the time to prepare a PR for my [simplification of Rx](https://github.com/fdietze/duality/blob/master/duality/shared/src/main/scala/rx/Rx.scala) which allows the implementation of some [bidirectional `Var` operators](https://github.com/fdietze/duality/blob/master/duality/shared/src/test/scala/rx/TransformedVarTests.scala), would it be merged? @Voltir?

The `Rx` simplification removes the `Internal` trait of `Rx` and makes its members `private[rx]` instead.

### PR DESCRIPTION
**Pull request by fdietze**
This PR Refactors `Core.scala`:
- Remove `Rx.Internal`
- `Rx.Dynamic` becomes invariant (Any Ideas how to fix this?)
- Extract trait for `Var`
- Rename `VarTuple` to `Var.Assignment`

Then splits up `Core.scala` to `Rx.scala`, `Var.scala`, `Ctx.scala` and `Obs.scala`.

This is all done to implement bidirectional `Var` operators `imap`, `zoom`, `mapRead`.

Be aware that this PR currently pulls in a dependency on [Monocle-cats](https://github.com/julien-truffaut/Monocle). This allows using lenses with the `zoom` operator. I'd rather implement lens support with a typeclass, but was not able to do that. Maybe this is possible with [`cats.Profunctor`](https://github.com/typelevel/cats/blob/master/core/src/main/scala/cats/arrow/Profunctor.scala).

fixes #93
fixes #84

### CODE REVIEW

File: scalarx/shared/src/main/scala/rx/Rx.scala [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+    now
+  }
+
+  /**
+    * Kills this [[Rx]]; stop listening for updates, and release all references
+    * to other [[Rx]]s. This lets the [[Rx]] be garbage-collected, since otherwise
+    * even when not-in-use it will continue to be referenced by the other [[Rx]]s
+    * it depends on.
+    */
+  private[rx] def kill(): Unit
```
Comment (Voltir) [inline scalarx/shared/src/main/scala/rx/Rx.scala]: I don't necessarily think this should be marked private. I have had occasions where some manual intervention on Rx behavior was necessary.

File: scalarx/shared/src/main/scala/rx/Rx.scala [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+
+  /**
+    * Force trigger/notifications of any downstream [[Rx]]s, without changing the current value
+    */
+  private[rx] def propagate(): Unit = Rx.doRecalc(downStream.toSet, observers)
```
Comment (Voltir) [inline scalarx/shared/src/main/scala/rx/Rx.scala]: Likewise I think propegate should be public

File: scalarx/shared/src/main/scala/rx/Rx.scala [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+
+  /**
+    * Force this [[Rx]] to recompute (whether or not any upstream [[Rx]]s
+    * changed) and propagate changes downstream. Does nothing if the [[Rx]]
+    * has been [[kill]]ed
+    */
+  private[rx] def recalc(): Unit
```
Comment (Voltir) [inline scalarx/shared/src/main/scala/rx/Rx.scala]: As well as recalc

File: scalarx/shared/src/test/scala/rx/BidirectionalVarOperators.scala [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+import monocle.macros.GenLens
+import utest._
+
+object TransformedVarTests extends TestSuite {
+  def tests = utest.Tests {
+    "isomorphic Var" - {
+      import Ctx.Owner.Unsafe._
+
+      val a: Var[Int] = Var(0)
+      val b: Var[String] = a.imap(_.toString)(_.toInt)
```
Comment (Voltir) [inline scalarx/shared/src/test/scala/rx/BidirectionalVarOperators.scala]: This is pretty neat :)

File: scalarx/shared/src/test/scala/rx/BidirectionalVarOperators.scala [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+          rawSelected().filter(list() contains _)
+        })
+      }
+
+      assert(selectedItem.now == Some(1))
+
+      selectedItem() = Some(4)
+      assert(selectedItem.now == None)
+
+      list.update(4 :: _)
```
Comment (Voltir) [inline scalarx/shared/src/test/scala/rx/BidirectionalVarOperators.scala]: This is magic.

File: scalarx/shared/src/test/scala/rx/BidirectionalVarOperators.scala [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+    }
+
+
+    "Composed Var" - {
+      import Ctx.Owner.Unsafe._
+
+      val list = Var(List(1, 2, 3))
+      val selectedItem = {
+        val rawSelected = Var[Option[Int]](Some(1))
+        new Var.Composed(rawSelected, Rx {
```
Comment (Voltir) [inline scalarx/shared/src/test/scala/rx/BidirectionalVarOperators.scala]: I'd love to see more examples / readme update of how Var.Composed works.

Comment (fdietze) [reply scalarx/shared/src/test/scala/rx/BidirectionalVarOperators.scala]: `mapRead` is essentially syntactic sugar for `Var.Composed`. This allows to construct safe consistency layers as described in #84.

**Comment by Voltir** [comment]
Slightly not related to this PR, what do you think about dropping 2.10 support?

**Comment by fdietze** [comment]
> what do you think about dropping 2.10 support?

I already did that in my fork. This simplifies the macro compatibility code.
Can we somehow find out how many users of scalarx_2.10 are out there?
I did a quick search, but the results are unclear:
https://github.com/search?l=&q=2.10+scalarx+language%3AScala&ref=advsearch&type=Code&utf8=%E2%9C%93
https://mvnrepository.com/artifact/com.lihaoyi/scalarx

**Comment by Voltir** [comment]
Er, I guess we can bump the version to 0.4.0 and drop 2.10 support at that version. I think the 0.3.3-SNAPSHOT changes already broke 2.10 compatibility unless we do a lot more work.

**Pull request merged**